cli-plugins/hooks/template.go

```diff
@@ -0,0 +1,127 @@
+package hooks
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+type HookType int
+
+const (
+	NextSteps = iota
+)
+
+// HookMessage represents a plugin hook response. Plugins
+// declaring support for CLI hooks need to print a json
+// representation of this type when their hook subcommand
+// is invoked.
+type HookMessage struct {
+	Type     HookType
+	Template string
+}
+
+// TemplateReplaceSubcommandName returns a hook template string
+// that will be replaced by the CLI subcommand being executed
+//
+// Example:
+//
+// "you ran the subcommand: " + TemplateReplaceSubcommandName()
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran the subcommand: run`
+func TemplateReplaceSubcommandName() string {
+	return hookTemplateCommandName
+}
+
+// TemplateReplaceFlagValue returns a hook template string
+// that will be replaced by the flags value.
+//
+// Example:
+//
+// "you ran a container named: " + TemplateReplaceFlagValue("name")
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran a container named: my-container`
+func TemplateReplaceFlagValue(flag string) string {
+	return fmt.Sprintf(hookTemplateFlagValue, flag)
+}
+
+// TemplateReplaceArg takes an index i and returns a hook
+// template string that the CLI will replace the template with
+// the ith argument, after processing the passed flags.
+//
+// Example:
+//
+// "run this image with `docker run " + TemplateReplaceArg(0) + "`"
+//
+// when being executed after the command:
+// `docker pull alpine`
+// will result in the message:
+// "Run this image with `docker run alpine`"
+func TemplateReplaceArg(i int) string {
+	return fmt.Sprintf(hookTemplateArg, strconv.Itoa(i))
+}
+
+const (
+	hookTemplateCommandName = "{docker-cli-plugin-hook-cmd-name}"
+	hookTemplateFlagValue   = "{docker-cli-plugin-hook-flag-%s}"
+	hookTemplateArg         = "{docker-cli-plugin-hook-arg-%s}"
+)
+
+var (
+	flagRegex = regexp.MustCompile(fmt.Sprintf(hookTemplateFlagValue, `([\w-]+)`))
+	argRegex  = regexp.MustCompile(fmt.Sprintf(hookTemplateArg, `([\w-]+)`))
+)
+
+func ProcessHook(b []byte, cmd *cobra.Command) (string, error) {
+	var hookMessageData HookMessage
+	err := json.Unmarshal(b, &hookMessageData)
+	if err != nil {
+		return "", err
+	}
+	return parseTemplate(hookMessageData.Template, cmd), nil
+}
+
+func parseTemplate(template string, cmd *cobra.Command) string {
+	ret := parseFlagValue(template, cmd)
+	ret = parseArg(ret, cmd)
+	ret = strings.ReplaceAll(ret, hookTemplateCommandName, cmd.Name())
+	return ret
+}
+
+func parseFlagValue(hookTemplate string, cmd *cobra.Command) string {
+	matches := flagRegex.FindAllStringSubmatch(hookTemplate, -1)
+	for _, m := range matches {
+		flagTemplate := m[0]
+		flag := cmd.Flag(m[1])
```

Comment:
maybe we should just make sure `m` is truly a length of 2 just in case.

Reply:
Not sure if that makes sense: the way the regex is constructed, we shouldn't get a match unless we also get a subexpression match. I'll confirm this is the case, and can add a comment in that case, but if that's the case I'd rather not add a check for something that is guaranteed not to happen.